applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_truss_element_3D2N.cpp

```diff
@@ -0,0 +1,262 @@
+    const double v2 = this->GetGeometry()[1].FastGetSolutionStepValue(DISPLACEMENT_Y);
+    const double w1 = this->GetGeometry()[0].FastGetSolutionStepValue(DISPLACEMENT_Z); 
+    const double w2 = this->GetGeometry()[1].FastGetSolutionStepValue(DISPLACEMENT_Z);
+
+    rDerivativeVector[0] = -1.0 * (u2 - u1 + dx) / l;
+    rDerivativeVector[1] = -1.0 * (v2 - v1 + dy) / l;
+    rDerivativeVector[2] = -1.0 * (w2 - w1 + dz) / l;
+    rDerivativeVector[3] = -1.0 * rDerivativeVector[0];
+    rDerivativeVector[4] = -1.0 * rDerivativeVector[1];
+    rDerivativeVector[5] = -1.0 * rDerivativeVector[2];
+   
+    KRATOS_CATCH("")
+}
+
+void AdjointFiniteDifferenceTrussElement::GetDerivativePreFactor(double& rDerivativePreFactor, const ProcessInfo& rCurrentProcessInfo)
+{
+    TracedStressType traced_stress_type = static_cast<TracedStressType>(this->GetValue(TRACED_STRESS_TYPE));
+
+    switch (traced_stress_type)
+    {
+        case TracedStressType::FX:
+        {
+            rDerivativePreFactor = this->CalculateDerivativePreFactorFX(rCurrentProcessInfo);
+            break;
+        }
+        case TracedStressType::PK2_TRUSS:
+        {
+            rDerivativePreFactor = this->CalculateDerivativePreFactorPK2(rCurrentProcessInfo);
+            break;
+        }
+        default:
+            KRATOS_ERROR << "Invalid stress type! Stress type not supported for this element!" << std::endl;
+    }
+}
+
+double AdjointFiniteDifferenceTrussElement::CalculateDerivativePreFactorFX(const ProcessInfo& rCurrentProcessInfo)
+{
+    const double numerical_limit = std::numeric_limits<double>::epsilon();
+    const double E = mpPrimalElement->GetProperties()[YOUNG_MODULUS];
+    const double A = mpPrimalElement->GetProperties()[CROSS_AREA];
+    const double L0 = CalculateReferenceLength();
+    const double l = CalculateCurrentLength();
+    double prestress = 0.00;
+    if (mpPrimalElement->GetProperties().Has(TRUSS_PRESTRESS_PK2))
+        prestress = mpPrimalElement->GetProperties()[TRUSS_PRESTRESS_PK2];
+    std::vector<Vector> GL_strain;  
+    mpPrimalElement->CalculateOnIntegrationPoints(GREEN_LAGRANGE_STRAIN_VECTOR, GL_strain, rCurrentProcessInfo);
+    const double GL_strain_X = GL_strain[0][0]; //one Gauss-Point result is enough due to constant strains.
+        
+    double derivative_pre_factor = A / L0 * (E * GL_strain_X + prestress + E * l * l / (L0 * L0));
+
+    KRATOS_ERROR_IF(std::abs(derivative_pre_factor) <= numerical_limit)
+        << "Derivative pre-factor of " << this->Id() << "~ 0" << std::endl;
+
+    return derivative_pre_factor;
+}
+
+double AdjointFiniteDifferenceTrussElement::CalculateDerivativePreFactorPK2(const ProcessInfo& rCurrentProcessInfo)
+{
+    const double numerical_limit = std::numeric_limits<double>::epsilon();
+    const double E = mpPrimalElement->GetProperties()[YOUNG_MODULUS];
+    const double l = CalculateCurrentLength();
+    const double L0 = CalculateReferenceLength();
```

Comment:
make l and L0 either both capital or small letters

Reply:
done